pkg/controller/direct/eventarc/channel_controller.go

```diff
@@ -0,0 +1,294 @@
+		gcpClient: eventarcClient,
+		id:        id,
+		desired:   obj,
+		reader:    reader,
+	}, nil
+}
+
+func (m *channelModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type channelAdapter struct {
+	gcpClient *gcp.Client
+	id        *krm.ChannelIdentity
+	desired   *krm.EventarcChannel
+	actual    *pb.Channel
+	reader    client.Reader
+}
+
+var _ directbase.Adapter = &channelAdapter{}
+
+func (a *channelAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting eventarc channel", "name", a.id)
+
+	req := &pb.GetChannelRequest{Name: a.id.String()}
+	actual, err := a.gcpClient.GetChannel(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting eventarc channel %q: %w", a.id.String(), err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *channelAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating eventarc channel", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	if err := a.normalizeReferenceFields(ctx); err != nil {
+		return err
+	}
+
+	desired := a.desired.DeepCopy()
+	resource := EventarcChannelSpec_ToProto(mapCtx, &desired.Spec)
+	resource.Name = a.id.String()
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &pb.CreateChannelRequest{
+		Parent:    a.id.Parent().String(),
+		Channel:   resource,
+		ChannelId: a.id.ID(),
+	}
+
+	op, err := a.gcpClient.CreateChannel(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating eventarc channel %q: %w", a.id.String(), err)
+	}
+
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("waiting for eventarc channel creation %q: %w", a.id.String(), err)
+	}
+	log.V(2).Info("successfully created eventarc channel", "name", a.id)
+
+	status := &krm.EventarcChannelStatus{}
+	status.ObservedState = EventarcChannelObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(created.Name)
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *channelAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating eventarc channel", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	if err := a.normalizeReferenceFields(ctx); err != nil {
+		return err
+	}
+
+	desired := a.desired.DeepCopy()
+	resource := EventarcChannelSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	paths := []string{}
+	if resource.Provider != a.actual.Provider {
```

Comment:
Could you use the standard comparison function like [this](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/direct/colab/runtimetemplate_controller.go#L248)?

Reply:
I thought compare proto was discouraged ?

Reply:
done.

Reply:
I meant i used compareProto. So this comment can be resolved.

Reply:
I'm not sure if there's a preferred implementation. The standard library can be problematic if a proto field is mislabeled. And sometimes the GCP API doesn't accept `updateMask` for child-level fields. Manually diffing provides more control in these cases.